src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -511,8 +511,24 @@ internal SynthesizedAttributeData SynthesizeTupleNamesAttributeOpt(TypeSymbol ty
             return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_TupleElementNamesAttribute__ctorTransformNames, args);
         }
 
-        private static class TupleNamesEncoder
+        internal static class TupleNamesEncoder
         {
+            public static ImmutableArray<string> Encode(TypeSymbol type)
+            {
+                var namesBuilder = ArrayBuilder<string>.GetInstance();
+                type.VisitType((t, builder, _ignore) => AddNames(t, builder), namesBuilder);
+                Debug.Assert(namesBuilder.Any());
+
+                // If none of the tuples have names, return a default array
+                if (namesBuilder.All(name => name == null))
+                {
+                    namesBuilder.Free();
+                    return default(ImmutableArray<string>);
+                }
```

Comment:
Consider extracting `VisitType` and `if (namesBuilder.All(...))` check to a helper: `TryGetNames(type, namesBuilder)`.